### ISSUE TITLE
din中seq_inputs的形状问题？

### ISSUE THREAD

**Issue by dominic-z** [ZiyaoGeng/Recommender-System-with-TF2.0#12]
`dense_inputs, sparse_inputs, seq_inputs, item_inputs = inputs`这一步，seq_inputs的shape是不是[batch_size, seq_length, field_num]格式的？我用数据测试了一下，
生成的形状好像是[batch_size, 40, 2]

**Comment by ZiyaoGeng** [comment]
是的，DIN代码我重新写了一遍

**Comment by dominic-z** [comment]
哇，感谢这么快的回复，我有个疑问是：
但是做embedding的时候使用的是`seq_embed = tf.concat([self.embed_seq_layers[i](seq_inputs[:, i]) for i in range(self.seq_len)], axis=-1)`，因为按照样例数据self.seq_len=2，但`seq_inputs[:, i]`只是每个样本的序列数据的第1位和第2位进行嵌入，输出格式就变成了`[batch_size, 2, 2*emb_dim]`，序列好像没有embedding全的样子?

**Comment by ZiyaoGeng** [comment]
并不是，首先我之前说错了，seq_inputs 的shape是[batch_size, file_num, seq_length], 你最后得到的seq_embed的shape是[batch_sze, seq_length, dim * 2],  我是将filed_num的embedding进行拼接。你可以调用下test_model， 然后打印下每个的shape看看

**Comment by dominic-z** [comment]
啊对还有个问题，就是在Attention_Layer的代码里`key_masks = tf.sign(tf.abs(tf.reduce_sum(k, axis=-1)))`这块，这个key_masks的计算方式我不太理解，如果k是`(None, seq_len, d * 2) `形状的，k是embedding后的结果，对其进行reduce_sum，即使原序列里某一个位置是padding项0，0也会参与embedding之后的结果不一定是一串0吧，例如如果输入序列是[0,1]，embedding结果可能是[[0.1,0.3],[0.3,0.1]]，这样计算可以找到正确的mask么？

**Comment by dominic-z** [comment]
> 并不是，首先我之前说错了，seq_inputs 的shape是[batch_size, file_num, seq_length], 你最后得到的seq_embed的shape是[batch_sze, seq_length, dim * 2], 我是将filed_num的embedding进行拼接。你可以调用下test_model， 然后打印下每个的shape看看

感谢回复
是的，我看到这个了，但test_model结果的正确是因为你覆写了summary方法，在summary方法里给定的输入是`seq_inputs = Input(shape=(self.seq_len, self.maxlen), dtype=tf.int32)`
但是在train.py的`feature_columns, behavior_list, train, val, test = create_amazon_electronic_dataset(file, embed_dim, maxlen)`这段代码里，`train`这个对象的数据包括`train_X, train_y`，而`seq_inputs`对应的就是`train_X[2]`，也就是`train[0][2]`，这个数据的形状是(2223958, 40, 2)的

**Comment by ZiyaoGeng** [comment]
这里确实写错了，以前的写法我是每个sample传入真实长度，然后发现其实不用，现在应该是通过seq_inputs来得到mask。

**Comment by ZiyaoGeng** [comment]
> > 并不是，首先我之前说错了，seq_inputs 的shape是[batch_size, file_num, seq_length], 你最后得到的seq_embed的shape是[batch_sze, seq_length, dim * 2], 我是将filed_num的embedding进行拼接。你可以调用下test_model， 然后打印下每个的shape看看
> 
> 感谢回复
> 是的，我看到这个了，但test_model结果的正确是因为你覆写了summary方法，在summary方法里给定的输入是`seq_inputs = Input(shape=(self.seq_len, self.maxlen), dtype=tf.int32)`
> 但是在train.py的`feature_columns, behavior_list, train, val, test = create_amazon_electronic_dataset(file, embed_dim, maxlen)`这段代码里，`train`这个对象的数据包括`train_X, train_y`，而`seq_inputs`对应的就是`train_X[2]`，也就是`train[0][2]`，这个数据的形状是(2223958, 40, 2)的

我再看看，重写的确实有点问题，感谢指出。

**Comment by dominic-z** [comment]
感谢回复，回复好及时，非常感谢，代码很棒，帮我打开了一些思路~我最近也在复现各种ctr的论文，加油！

**Comment by wanglc2008** [comment]
作者赶紧修复一下吧，代码写的很不错，受教了

**Comment by ZiyaoGeng** [comment]
已更正
